### PR TITLE
Module guidance

### PR DESCRIPTION
**Pull request by mhegazy**
Supersedes  #85

Add content in http://stackoverflow.com/questions/30357634/how-do-i-use-namespaces-with-typescript-external-modules to modules.md

### CODE REVIEW

**Comment by sandersn** [comment]
Looks good to me.

File: pages/Modules.md [hunk 1]
```diff
@@ -498,3 +498,247 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
 import * as URL from "url";
 var myUrl = URL.parse("http://www.typescriptlang.org");
 ```
+
+# Guidance for structuring modules
+
+## Export as close to top-level as possible
+
+### If you're only exporting a single `class` or `function`, use `export default`
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: consider using

File: pages/Modules.md [hunk 2]
```diff
@@ -498,3 +498,247 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
 import * as URL from "url";
 var myUrl = URL.parse("http://www.typescriptlang.org");
 ```
+
+# Guidance for structuring modules
+
+## Export as close to top-level as possible
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: What does this mean?

File: pages/Modules.md [hunk 3]
```diff
@@ -498,3 +498,247 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
...
+
+import { Calculator } from "./Calculator";
+
+
+class ProgrammerCalculator extends Calculator {
+    static digits = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"];
+
+    constructor(public base: number) {
+        super();
+        if (base > ProgrammerCalculator.digits.length) {
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: Check for `0 <= base` as well

File: pages/Modules.md [hunk 4]
```diff
@@ -498,3 +498,247 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
...
+            throw new Error("base has to be at most 16");
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: Within 0 to 16 inclusive.

File: pages/Modules.md (hunk 2 above)
Comment (mhegazy) [inline pages/Modules.md]: do not export a class with a static helper, export a class and a function.. similarly do not export a namespace.

File: pages/Modules.md [hunk 5]
```diff
@@ -498,3 +498,247 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
...
+Now to extend this to add support for input with numbers in bases other than 10, let's create `ProgrammerCalculator.ts`
+
+#### ProgrammerCalculator.ts
+
+```ts
+
+import { Calculator } from "./Calculator";
+
+
+class ProgrammerCalculator extends Calculator {
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: Why didn't you default-export this one?

Comment (DanielRosenwasser) [inline pages/Modules.md]: Never mind. We should amend this when we support the new export forms

File: pages/Modules.md (hunk 2 above)
Comment (DanielRosenwasser) [inline pages/Modules.md]: I've elaborated on it a little. Let me know what you think of the changes.

File: pages/Modules.md [hunk 6]
```diff
@@ -498,3 +498,262 @@ Now we can `/// <reference>` `node.d.ts` and then load the modules using `import
...
+
+Static methods on an exported class have a similar problem - the class itself adds a layer of nesting.
+Unless it increases expressivity or intent in a clearly useful way, consider simply exporting a helper function.
+
+### If you're only exporting a single `class` or `function`, use `export default`
+
+Just as "exporting near the top-level" reduces friction on your module's consumers, so does introducing a default export.
+If a module's primary purpose is to house one specific export, then you should consider exporting it as a default export.
+This makes both importing and actually using the import a little easier.
+For example:
```
Comment (DanielRosenwasser) [inline pages/Modules.md]: I've also elaborated here. Take a look @mhegazy

File: pages/Modules.md (hunk 2 above)
Comment (mhegazy) [inline pages/Modules.md]: looks good. thanks!

File: pages/Modules.md (hunk 3 above)
Comment (mhegazy) [inline pages/Modules.md]: done

File: pages/Modules.md (hunk 4 above)
Comment (mhegazy) [inline pages/Modules.md]: done.

**Pull request merged**